File: app/src/main/java/com/typingsolutions/kore/common/DatabaseConnection.java
package com.typingsolutions.kore.common;

import android.content.Context;
import android.util.Log;
import net.sqlcipher.database.SQLiteDatabase;
import net.sqlcipher.database.SQLiteDatabaseHook;
import net.sqlcipher.database.SQLiteOpenHelper;

public class DatabaseConnection extends SQLiteOpenHelper {
  private static final String NAME = "database.db";
  private static final int VERSION = 1;


  private String mPassword;
  private int mPim;

  DatabaseConnection(Context context, String password, int pim) {
    super(context, NAME, null, VERSION, getDatabaseHook(pim));
    mPassword = password;
    mPim = pim;

  }

  @Override
  public void onCreate(SQLiteDatabase sqLiteDatabase) {

  }

  @Override
  public void onUpgrade(SQLiteDatabase sqLiteDatabase, int i, int i1) {

  }

  @Override
  public synchronized void close() {
    super.close();

    mPim = 0;

    // TODO: clear string pool
    mPassword = null;
  }

  public synchronized SQLiteDatabase getWritableDatabase() {
    if (mPassword == null || mPassword.length() == 0) return null;

    return super.getWritableDatabase(mPassword);
  }

  private static SQLiteDatabaseHook getDatabaseHook(int pim) {
    final String iterations = Integer.toString(pim);
    Log.d(DatabaseConnection.class.getSimpleName(), iterations);

    return new SQLiteDatabaseHook() {
      @Override
      public void preKey(SQLiteDatabase sqLiteDatabase) {
      }

      @Override
      public void postKey(SQLiteDatabase sqLiteDatabase) {
        sqLiteDatabase.rawExecSQL(String.format("PRAGMA kdf_iter = %s", iterations));
      }
    };
  }
}

File: app/src/main/java/com/typingsolutions/kore/common/DatabaseConnection.kt
package com.typingsolutions.kore.common

import android.content.Context
import android.util.Log
import com.typingsolutions.kore.common.data.History
import com.typingsolutions.kore.common.data.Password
import net.sqlcipher.database.SQLiteDatabase
import net.sqlcipher.database.SQLiteDatabaseHook
import net.sqlcipher.database.SQLiteOpenHelper

class DatabaseConnection internal constructor(context: Context, private var mPassword: String?, private var mPim: Int) : SQLiteOpenHelper(context, DatabaseConnection.NAME, null, DatabaseConnection.VERSION, DatabaseConnection.getDatabaseHook(mPim)) {

    override fun onCreate(sqLiteDatabase: SQLiteDatabase) {
        sqLiteDatabase.execSQL(Password.CreateString)
        sqLiteDatabase.execSQL(History.CreateString)
    }

    override fun onUpgrade(sqLiteDatabase: SQLiteDatabase, i: Int, i1: Int) {

    }

    @Synchronized override fun close() {
        super.close()

        mPim = 0

        // TODO: clear string pool
        mPassword = null
    }

    val writableDatabase: SQLiteDatabase?
        @Synchronized get() {
            if (mPassword == null || mPassword?.length == 0) return null

            return super.getWritableDatabase(mPassword)
        }

    companion object {
        val NAME = "kore.db"
        private val VERSION = 1

        private fun getDatabaseHook(pim: Int): SQLiteDatabaseHook {
            val iterations = Integer.toString(pim)
            Log.d(DatabaseConnection::class.java.simpleName, iterations)

            return object : SQLiteDatabaseHook {
                override fun preKey(sqLiteDatabase: SQLiteDatabase) {}

                override fun postKey(sqLiteDatabase: SQLiteDatabase) {
                    sqLiteDatabase.rawExecSQL(String.format("PRAGMA kdf_iter = %s", iterations))
                }
            }
        }
    }
}
